### DISCUSSION TITLE
Enforce rule of three or rule of five for future ESBMC extensions?

### DISCUSSION THREAD

**Discussion by kunjsong01** [esbmc/esbmc#603] [General]
Shall we enforce the C++ rule of three for future ESBMC development? If smart pointers are used, perhaps escalate it to rule of five? 

And explicitly add default/delete copy constructors and copy assignment operators in the class, so that the compiler would report error when a shallow copy is about to happen in the code changes. This would save us lots of debugging time, and prevent unexpected failure due to shallow copy.

**Comment by rafaelsamenezes** [comment]
Is there any static analysis tool that can help us to ensure this rule?

**Reply by kunjsong01** [reply]
clang-tidy seems to have check for this rule. (https://sbmueller.github.io/posts/clang/)

**Reply by rafaelsamenezes** [reply]
> clang-tidy seems to have check for this rule. (https://sbmueller.github.io/posts/clang/)

That would nice, we already have support for it in the CI.

- https://github.com/esbmc/esbmc/issues/561
- https://github.com/esbmc/esbmc/issues/224

**Comment by lucasccordeiro** [comment]
> Shall we enforce the C++ rule of three for future ESBMC development? 
> If smart pointers are used, perhaps escalate it to rule of five?

Yes. Yes. I agree with these suggestions :-)

**Comment by fbrausse** [comment]
> Shall we enforce the C++ rule of three for future ESBMC development? If smart pointers are used, perhaps escalate it to rule of five?

Definitely yes!

> And explicitly add default/delete copy constructors and copy assignment operators in the class, so that the compiler would report error when a shallow copy is about to happen in the code changes. This would save us lots of debugging time, and prevent unexpected failure due to shallow copy.

Can you elaborate how adding defaulted constructors/operators help in that regard? Deleted ones and defaulted copy when move is defined, sure, because they are required for the correct behaviour. But for example
```c++
struct A {
  A(const A &) = default;
  A(A &&) = default;
  A & operator=(const A &) = default;
  A & operator=(A &&) = default;
};
```
seems pointless.

**Reply by fbrausse** [reply]
This is the version of the "rule of 5" I usually try to adhere to:
```c++
struct A {
  A(const A &);
  A(A &&);
  friend void swap(A &a, A &b) { using std::swap; /* swap() all members */ }
  A & operator=(A a /* by-value */) { swap(*this, a); return *this; }
};
```
It keeps working if either constructor is removed, too + provides `swap()` more or less for free.
